src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/AuthenticatorFactoryInterface.php

```diff
@@ -0,0 +1,29 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\SecurityBundle\DependencyInjection\Security\Factory;
+
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ *
+ * @experimental in 5.1
+ */
+interface AuthenticatorFactoryInterface
+{
+    /**
+     * Creates the authenticator service(s) for the provided configuration.
+     *
+     * @return string|string[] The authenticator service ID(s) to be used by the firewall
+     */
+    public function createAuthenticator(ContainerBuilder $container, string $id, array $config, string $userProviderId);
```

Comment:
If this "authenticator factory" can only create authenticators, we could rename this method as `create()`:

```php
$authenticatorFactory->createAuthenticator('...');

$authenticatorFactory->create('...');
```

Reply:
This makes it impossible to implement the interface on the same class than the SecurityFactoryInterface of the old system, while this is how things are done currently (see AnonymousFactory for instance).

Reply:
I see ... you are right ... I always forgot about the micro-factories that Symfony uses in some places 😅